---
title: "Using Mixscale for Perturb-seq data"
author: "Longda Jiang"
date: "2024-01-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

In this tutorial, we will describe an R package "Mixscale" for analyzing Perturb-seq data. Mixscale contains functions designed to tackle the following tasks:\
1. calculate 'Mixscale scores' for cells that receives the same perturbation to quantify the heterogeneity in perturbation strength \
2. perform a scoring-based weighted differential expression (DE) tests to identify DE genes for each perturbation \
3. perform different levels of decomposition analysis to identify correlated perturbations and group them into a program \
4. perform a PCA-based permutation test to extract shared genes for the perturbation programs (program gene signature) \
5. identify shared and unique signature between two relevant programs \
6. perform gene set enrichment tests using the program signature for new datasets \
7. perform module score analyses using the program signature to quantify the program activity in new datasets \
\
The tutorial is divided into two sections. The first section will describe task 1 to 5 using a public Perturb-seq dataset from the Weissman Lab ([Jost et al 2020](https://www.nature.com/articles/s41587-019-0387-5)), which can be downloaded from [GSE132080](https://0-www-ncbi-nlm-nih-gov.brum.beds.ac.uk/geo/query/acc.cgi?acc=GSE132080). The second section will describe task 6 and 7 using the pathway gene lists generated from our study (available at [Zenodo](not_inserted_yet)) and an interferon-beta stimulated human PBMCs dataset (ifnb) from [Kang et al 2017](https://www.nature.com/articles/nbt.4042). This ifnb dataset is available via the SeuratData package (see section 2 below).


### load the packages
```{r load_package, message=FALSE, warning=FALSE}
options(Seurat.object.assay.version = 'v3')

library(Seurat)
library(ggridges)
library(stringr)
library(Mixscale)
```

## Section 1. 
In this section we will focus on how to use Mixscale to analyze Perturb-seq data. 

### 0. load the demo data 
The demo dataset from [Jost et al 2020](https://www.nature.com/articles/s41587-019-0387-5) contains CRISPRi Perturb-seq data targeting 25 key genes involved in essential cell biological processes. We will load in the count matrix to create a Seurat object and append the provided meta data to it. \
One special feature of this dataset is that, for each perturbation target gene, there are five different gRNAs designed to target it. One of the gRNA has the perfectly matched sequence for the target region (labelled with "_00"), while the others contain 1~3 nucleotide mismatches so that their perturbation stength is "titrated". We will treat the cells that have the same target gene as the same group in our downstream analyses.

```{r load_data, message=FALSE, warning=FALSE, cache=TRUE}
# load the count matrix
ct_mat = ReadMtx(mtx = "/Users/uqljian5/Desktop/Lab_stuffs_NYGC/Paper_information_gathering/results/GSE132080/GSE132080_10X_matrix.mtx", 
                 cells = "/Users/uqljian5/Desktop/Lab_stuffs_NYGC/Paper_information_gathering/results/GSE132080/GSE132080_10X_barcodes.tsv", 
                 features = "/Users/uqljian5/Desktop/Lab_stuffs_NYGC/Paper_information_gathering/results/GSE132080/GSE132080_10X_genes.tsv")
# load the meta_data
meta_data = read.csv("/Users/uqljian5/Desktop/Lab_stuffs_NYGC/Paper_information_gathering/results/GSE132080/GSE132080_cell_identities.csv")
rownames(meta_data) = meta_data$cell_barcode

# create a seurat object 
seurat_obj = CreateSeuratObject(counts = ct_mat, meta.data = meta_data)
rm(ct_mat, meta_data)

# retrieve the guide information for each cell
txt = seurat_obj$guide_identity
txt2 = str_extract(txt, "^[^_]+")
txt3 = gsub(pattern = "^[^_]+_", replacement = "", txt)
seurat_obj[['gene']] = txt2
seurat_obj[['gRNA_name']] = txt3

# remove ambiguous cells 
seurat_obj = subset(seurat_obj, subset = number_of_cells == 1)  # 19594 cells remain
seurat_obj = subset(seurat_obj, subset = guide_identity != '*')  # 19587 cells remain

seurat_obj
```

### 1. Pre-processing and calculating the Mixscale score
We will first run standard pre-processing (normalization, find variable features, etc) for the dataset. Then, we will follow the standard [Mixscape analysis](https://satijalab.org/seurat/articles/mixscape_vignette) to calculate local perturbation signatures that mitigate confounding effects. Briefly speaking, for each cell we will search for its 20 nearest neighbors from the non-targeted (NT) cells, and then remove all technical variation so that perturbation-specific effect can be revealed.

```{r standard_process, echo=TRUE, message=FALSE, warning=FALSE, cache=TRUE}
# standard pre-processing
seurat_obj = NormalizeData(seurat_obj)
seurat_obj = FindVariableFeatures(seurat_obj)
seurat_obj = ScaleData(seurat_obj)
seurat_obj = RunPCA(seurat_obj)

# calculate Perturbation signatures 
seurat_obj <- CalcPerturbSig(
    object = seurat_obj, 
    assay = "RNA", 
    slot = "data", 
    gd.class ="gene", 
    nt.cell.class = "neg", 
    reduction = "pca", 
    ndims = 40, 
    num.neighbors = 20, 
    new.assay.name = "PRTB")

```

Now we will calculate the Mixscale scores for each cell within each perturbation group. 
```{r scoring, echo=TRUE, message=FALSE, warning=FALSE, cache=TRUE}
# Mixscale
seurat_obj = RunMixscale(
    object = seurat_obj, 
    assay = "PRTB", 
    slot = "scale.data", 
    labels = "gene", 
    nt.class.name = "neg", 
    min.de.genes = 5, 
    logfc.threshold = 0.2,
    de.assay = "RNA",
    max.de.genes = 100, 
    prtb.type = "P", new.class.name = "mixscale_id", fine.mode = F)

```


### 2. Visualizations for the scores
We will now use some plotting functions to explore the perturbation scores that we just calculated. 

```{r, echo=TRUE, message=FALSE, warning=FALSE, cache=TRUE}
# a. Check the distribution of the scores for the first 10 perturbations
Mixscale_RidgePlot(object = seurat_obj, 
                   nt.class.name = "neg", 
                    PRTB = unique(seurat_obj$gene)[unique(seurat_obj$gene) != "neg"][1:10], 
                    facet_wrap = "gene", facet_scale = "fixed")

# b. Check if the scores correlate with the expression level of the target gene itself
Mixscale_ScatterPlot(object = seurat_obj, nt.class.name = "neg", 
                     PRTB = unique(seurat_obj$gene)[unique(seurat_obj$gene) != "neg"][1:10], 
                      facet_wrap = "gene", facet_scale = "free_y", nbin = 10)
```

### 3. Differential expression (DE) analysis 
After calculating the scores, we can use the scores to enhance the statistical power of DE analysis by using them as a "weights" in the regression model. Briefly speaking, instead of coding the NT cells as 0 and the targeted cells as 1, we used the standardized scores to code the targeted cells in the regression, so that cells with stronger perturbation strength will have higher "weights" and vice versa. 

```{r, echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE}
# run score-based weighted DE test for 12 selected perturbations. It will return a list of data frames (one for each perturbation)
de_res = Run_wtDE(object = seurat_obj, assay = "RNA", slot = "counts",
                  labels = "gene", nt.class.name = "neg", 
                  logfc.threshold = 0.1)

# have a quick look at the DE results
head(de_res[[1]])

```

We can now explore the top DE genes for each perturbations using the customized DoHeatmap function, where cells are ordered by Mixscale scores. 

```{r, echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE}
# heatmap for the top DE genes
Mixscale_DoHeatmap(object = seurat_obj, PRTB = "POLR2H", 
                   slct_condition = "con1", 
                   nt.class.name = "neg",
                   labels = "gene",
                   slct_features = rownames(de_res[["POLR2H"]][order(de_res[["POLR2H"]]$p_weight), ])[1:20]) + NoLegend()

# similar heatmap for the top DE genes, but this time the cells are divided based on gRNA identity using slct_ident
Mixscale_DoHeatmap(object = seurat_obj, PRTB = "GATA1", 
                   slct_condition = "con1", 
                   nt.class.name = "neg",
                   labels = "gene",
                   slct_features = rownames(de_res[["GATA1"]][order(de_res[["GATA1"]]$p_weight), ])[1:20], 
                   slct_ident = "gRNA_name") + NoLegend()
```

### 5. Decomposition analyses to identify correlated perturbations 
#### 5.1 Hierarchical clustering 
In this section we will perform mainly two types of decomposition analyses for our DE results. The first is a hierarchical clustering analysis based on  ([MinMax](https://www.sciencedirect.com/science/article/abs/pii/S0031320314000338)). We will apply it to the DE Z-score matrix of our DE results. 

```{r, echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE}
# get the Z-score matrix (a list of matrices will be returned, with each matrix correspond to a cell type)
# as for the selection of rows (features), we will use the union set of the top 100 DE genes from each column
DEG_mat_main = get_DE_mat(de_res, p_threshold = 0.05/30000, fc_threshold = 0.2, num_top_DEG = 100)

# slightly clean up the matrices by removing columns with not enough significant DEGs and replace all the NAs with 0
DEG_mat = prune_DE_mat(DEG_mat_main, mask_target = T, min_sig_DEG = 5)

# an empty list to store all the gene sets (a repository of gene sets)
go_db = list()

# get the cell type (in this example there is only one cell type)
celltype_list = names(DEG_mat)

# loop through all the cell types to perform Minmax hierarchical clustering
for(i in 1:length(celltype_list)){
    CELLTYPE = celltype_list[i]
    tmp=DEG_mat[[CELLTYPE]]
    
    # run hierarchical clustering using Minmax. Other standard hclust methods are also supported. 
    # dist_thres defines the height (= 1 - dist_thres) that is used to cut the hclust tree. 
    # a lower value indicates a more stringent threshold to define clusters.
    res = DEhclust(mat = tmp, cor_method = "pearson", hclust_method = "minmax", dist_thres = 0.4)
    
    # get_sig_genes_DEhclust() is a wrapper function for PCApermtest() and get_sig_genes(), which will
    # perform a PCA-based permutation test and extract the top shared DE genes across the perturbations 
    # in the same cluster
    sig_genes = get_sig_genes_DEhclust(obj = res, row_filtering_pval = 0.05)
    
    # store the extracted top genes as the cluster signature into go_db
    for(CLUSTER in names(sig_genes)){
        if(length(sig_genes[[CLUSTER]]$sig_genes$upDEGs) >= 10){
            go_db[[paste0(CELLTYPE, "_", CLUSTER, "_upDEGs")]] = sig_genes[[CLUSTER]]$sig_genes$upDEGs
        } 
        if(length(sig_genes[[CLUSTER]]$sig_genes$downDEGs) >= 10){
            go_db[[paste0(CELLTYPE, "_", CLUSTER,"_downDEGs")]] = sig_genes[[CLUSTER]]$sig_genes$downDEGs
        } 
    }
    
}

# check the clustering results 
res$cluster_assignment

# generate a correlation matrix plot based on the clustering results 
col3 = rev(brewer.pal(11,"RdBu"))
heatmap.2(cor(tmp), 
          Rowv = as.dendrogram(res$hclust), 
          Colv = as.dendrogram(res$hclust), 
          dendrogram = "none", 
          col = col3)
```

we can also use the following function to generate Z-score heatmap for each perturbation cluster and save them to a user defined directory 
```{r, echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE, eval=FALSE}
DE_heatmap(obj = res, sig_genes = sig_genes, type = "hclust", direction = "both", top_n = 30, 
        output_path = "output_path/", 
        prefix = CELLTYPE)
```

#### 5.2 MultiCCA analysis
In our [paper](not_insert_yet), we introduced a novel approach for identifying correlated perturbations both within and between various matrices. This method is applicable when DE Z-scores are organized into a list of Z-score matrices, with each matrix corresponding to a cell type. This is especially useful when multiple cell types/lines are used in a Perturb-seq experiment. The demo dataset only contains one cell type, so we will randomly divide the Z-score matrix above into a list of 3 matrices, and use them to test our MultiCCA method.

```{r, echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE}
# first we randomly divide the DEG_mat_main into a list of 3 matrices 
set.seed(100)
DEG_mat_main2 = list(type1 = DEG_mat_main[, c(1, sample(2:26)[1:8])],
                     type2 = DEG_mat_main[, c(1, sample(2:26)[9:16])], 
                     type2 = DEG_mat_main[, c(1, sample(2:26)[17:25])] )

# clean up the matrices 
DEG_mat = prune_DE_mat(DEG_mat_main2, min_sig_DEG = 5, center = T)

#  run MultiCCA 
res = DEmultiCCA(mat_list = DEG_mat, cor_coef_thres = 0.8, mean_cor_thres = 0.2, max_k = 3, standardize = F)  

# get_sig_genes_DEmultiCCA is a wrapper function for PCApermtest() and get_sig_genes() for DEmultiCCA object.
sig_genes = get_sig_genes_DEmultiCCA(res, row_filtering_pval = 0.05)

# store the gene signatures to the go-term repo
for(PROGRAM in names(sig_genes)){
    if(length(sig_genes[[PROGRAM]]$sig_genes$upDEGs) >= 10){
        go_db[[paste0("MultiCCA_", PROGRAM, "_upDEGs")]] = sig_genes[[PROGRAM]]$sig_genes$upDEGs
    } 
    if(length(sig_genes[[PROGRAM]]$sig_genes$downDEGs) >= 10){
        go_db[[paste0("MultiCCA_", PROGRAM, "_downDEGs")]] = sig_genes[[PROGRAM]]$sig_genes$downDEGs
    } 
}

```

Again, we can use the following function to generate Z-score heatmap for each perturbation program and save them to a user defined directory 
```{r, echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE, eval=FALSE}
DE_heatmap(obj = res, sig_genes = sig_genes, 
           type = "multiCCA", direction = "both", 
           top_n = 30, labRow = T, 
           output_path = "output_path/", 
           prefix = "MultiCCA")
```








